var/spack/repos/builtin/packages/openfoam-com/package.py

```diff
@@ -415,6 +419,15 @@ def patch(self):
             posix=join_path('etc', 'bashrc'),
             cshell=join_path('etc', 'cshrc'))
 
+        # This is a workaround for concretizer limitation (2017-09-26)
+        if self.spec['mpi'].name == 'openmpi':
```

Comment:
i don't understand why this is needed. Either `conflicts()` is not working above or you use this incorrectly.

@alalazo any ideas?

Reply:
`conflicts()` does not work when the mpi type is not explicitly specified. openfoam itself only `depends_on('mpi')` so we might get openmpi, intelmpi, mpich whatever. If I examine the output from `spack spec openfoam-com`  I automatically get an `openmpi~thread_multiple` coming in via fftw:

     ^fftw@3.3.6-pl2%gcc@4.8+double+float+long_double+mpi~openmp~pfft_patches~quad arch=linux-opensuse42-x86_64 
        ^openmpi@3.0.0%gcc@4.8~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=linux-opensuse42-x86_64 

It doesn't really seem that the `conflicts()` descends below the first level of depends.

Reply:
> conflicts() does not work when the mpi type is not explicitly specified.

you don't use it right, you should do 
```
    conflicts('^openmpi~thread_multiple', when='@x.y.z:')
```
and throw away this run-time check.

p.s. tested this on `hdf5`, works as expected:

```
$ spack spec hdf5
Input spec
--------------------------------
hdf5

Normalized
--------------------------------
hdf5
    ^zlib@1.1.2:

Concretized
--------------------------------
==> Error: Conflicts in concretized spec "hdf5@1.10.1%clang@9.0.0-apple+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=darwin-sierra-x86_64 /qp5la73"

List of matching conflicts for spec:

    hdf5@1.10.1%clang@9.0.0-apple+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=darwin-sierra-x86_64
        ^openmpi@3.0.0%clang@9.0.0-apple~cuda fabrics= ~java schedulers= ~sqlite3~thread_multiple+vt arch=darwin-sierra-x86_64
            ^hwloc@1.11.7%clang@9.0.0-apple~cuda+libxml2~pci arch=darwin-sierra-x86_64
                ^libxml2@2.9.4%clang@9.0.0-apple~python arch=darwin-sierra-x86_64
                    ^pkg-config@0.29.2%clang@9.0.0-apple+internal_glib arch=darwin-sierra-x86_64
                    ^xz@5.2.3%clang@9.0.0-apple arch=darwin-sierra-x86_64
                    ^zlib@1.2.11%clang@9.0.0-apple+pic+shared arch=darwin-sierra-x86_64

1. "^openmpi~thread_multiple" conflicts with "hdf5@1.8.19:"
```
vs
```
$ spack spec hdf5+mpi ^openmpi+thread_multiple
Input spec
--------------------------------
hdf5+mpi
    ^openmpi+thread_multiple

Normalized
--------------------------------
hdf5+mpi
    ^openmpi+thread_multiple
        ^hwloc
            ^pkg-config
    ^zlib@1.1.2:

Concretized
--------------------------------
hdf5@1.10.1%clang@9.0.0-apple+cxx~debug+fortran+mpi+pic+shared~szip~threadsafe arch=darwin-sierra-x86_64
    ^openmpi@3.0.0%clang@9.0.0-apple~cuda fabrics= ~java schedulers= ~sqlite3+thread_multiple+vt arch=darwin-sierra-x86_64
        ^hwloc@1.11.7%clang@9.0.0-apple~cuda+libxml2~pci arch=darwin-sierra-x86_64
            ^libxml2@2.9.4%clang@9.0.0-apple~python arch=darwin-sierra-x86_64
                ^pkg-config@0.29.2%clang@9.0.0-apple+internal_glib arch=darwin-sierra-x86_64
                ^xz@5.2.3%clang@9.0.0-apple arch=darwin-sierra-x86_64
                ^zlib@1.2.11%clang@9.0.0-apple+pic+shared arch=darwin-sierra-x86_64
```

Reply:
But this will never give me an automatic selection of `+thread_multiple`, which is what I want.

Reply:
> But this will never give me an automatic selection of +thread_multiple, which is what I want.

your current solution does not give it to you either, right? I mentioned it in the issue already, that to my knowledge what you want is currently impossible. 

But this is irrelevant, the PR in its current state still uses conflicts incorrectly.